tools/clang/lib/Sema/SemaHLSL.cpp

```diff
@@ -15790,41 +15888,71 @@ clang::FunctionDecl *ValidateNoRecursion(clang::Sema *self,
   // on functions that are unreachable (as an early form of dead code
   // elimination).
   if (FD) {
-    hlsl::CallGraphWithRecurseGuard CG;
-    CG.BuildForEntry(FD);
-    return CG.CheckRecursion(FD);
+    HLSLExternalSource *hlslSource = HLSLExternalSource::FromSema(self);
+    hlslSource->BuildCallGraphForEntry(FD);
+    return hlslSource->CheckForCallGraphRecursion(FD);
   }
   return nullptr;
 }
 
+// There is an assumption being made that this function will only be called
+// on library shader stages.Thus, there is no entry point, only exported
+// functions
 void ValidateNoRecursionInTranslationUnit(clang::Sema *self) {
   std::set<FunctionDecl *> FDecls;
   std::vector<FunctionDecl *> FDeclsVec;
+  std::vector<FunctionDecl *> AllExportedFDecls;
+
   for (auto decl : self->getASTContext().getTranslationUnitDecl()->decls()) {
-    // TODO: improve condition so that only exported functions are checked,
-    // instead of all functions. Issue: #5857
     if (FunctionDecl *FD = dyn_cast<FunctionDecl>(decl)) {
-      // returns the first recursive function declaration detected
-      // from this function declaration FD, and determines whether
-      // the recursion was detected in the patch-constant function
-      FunctionDecl *pResult = ValidateNoRecursion(self, FD);
-      // if there is a function that was detected to be recursive,
-      // then make sure it is saved for later to emit a diagnostic
-      if (pResult) {
-        FDecls.insert(pResult);
-        FDeclsVec.push_back(pResult);
+      if (!FD->hasBody() ||
+          !IsExported(self, FD, getDefaultLinkageExternal(self))) {
+        continue;
+      }
+      AllExportedFDecls.push_back(FD);
+    } else if (DeclContext *declContext = dyn_cast<DeclContext>(decl)) {
+      // Loop through all declarations inside the class or namespace
+      for (DeclContext::decl_iterator j = declContext->decls_begin(),
+                                      e = declContext->decls_end();
+           j != e; ++j) {
+        // Check if the declaration is a function declaration
+        if (FunctionDecl *functionDecl = dyn_cast<FunctionDecl>(*j)) {
+          if (!functionDecl->hasBody() ||
+              !IsExported(self, functionDecl,
+                          getDefaultLinkageExternal(self))) {
+            continue;
+          }
+          AllExportedFDecls.push_back(functionDecl);
+        }
       }
     }
   }
 
+  for (auto FD : AllExportedFDecls) {
+    // returns the first recursive function declaration detected
+    // from this function declaration FD, and determines whether
+    // the recursion was detected in the patch-constant function
+    if (!FD->hasBody() ||
+        !IsExported(self, FD, getDefaultLinkageExternal(self))) {
+      continue;
```

Comment:
Presumably, this was already checked when adding to `AllExportedFDecls`, so you don't need to do it here again.